### PR TITLE
✨ 51 adding form components

### PR DESCRIPTION
**Pull request by langehm**
Issues #51

### CODE REVIEW

**Comment by langehm** [comment]
The search form does need some logic for calculating the current available options. Should we use some debounce logic to reduce computation?
```ts
const debounce = (func: (...args: any[]) => any, timeout: number = 500) => {
  let timer: number;
  return (...args: any[]) => {
    clearTimeout(timer);
    timer = setTimeout(() => {
      func.apply(this, args);
    }, timeout);
  };
};
```

**Comment by langehm** [comment]
Suffix does not seem to work in the patternlab either ...
grafik

Prefix does work.

**Comment by langehm** [comment]
some forms are missing ...

**Comment by langehm** [comment]
> Suffix does not seem to work in the patternlab either ... grafik
> 
> Prefix does work.

Works only with icons

File: src/components/Form/MucCheckbox.stories.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import MucCheckbox from "./MucCheckbox.vue";
+
+export default {
+  component: MucCheckbox,
+  title: "Input/MucCheckBox",
```
Comment (FabianWilms) [inline src/components/Form/MucCheckbox.stories.ts:5]:
```suggestion
  title: "Forms/MucCheckBox",
```

I would prefer a "Forms"-Category

File: src/components/Form/MucCheckbox.stories.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  tags: ["autodocs"],
+  parameters: {
+    docs: {
+      description: {
+        component: `The muc-text-area component is an input component for larger texts.
+
+[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=elements-textarea)
```
Comment (FabianWilms) [inline src/components/Form/MucCheckbox.stories.ts:12]: Wrong comment and link?

File: src/components/Form/MucCheckbox.vue [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+      class="m-label m-checkboxes__label"
+      for="checkbox-example-"
+    >
+      {{ label }}
+    </label>
+  </div>
+</template>
+
+<script setup lang="ts">
+const checkBoxValue = defineModel({ default: false });
```
Comment (FabianWilms) [inline src/components/Form/MucCheckbox.vue:22]: In Vue I would prefer to have this a model-value and not an internal represantation. This way I can't change the state of the checkbox from outside.

File: src/components/Form/MucCheckboxGroup.stories.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import MucCheckboxGroup from "./MucCheckboxGroup.vue";
+
+export default {
+  component: MucCheckboxGroup,
+  title: "Input/MucCheckboxGroup",
```
Comment (FabianWilms) [inline src/components/Form/MucCheckboxGroup.stories.ts:5]:
```suggestion
  title: "Forms/MucCheckboxGroup",
```

File: src/components/Form/MucCheckboxGroup.stories.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  tags: ["autodocs"],
+  parameters: {
+    docs: {
+      description: {
+        component: `The muc-text-area component is an input component for larger texts.
+
+[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=elements-textarea)
```
Comment (FabianWilms) [inline src/components/Form/MucCheckboxGroup.stories.ts:12]: Wrong link and comment?

File: src/components/Form/MucInput.stories.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
+import MucInput from "./MucInput.vue";
+
+export default {
+  component: MucInput,
+  title: "Input/MucInput",
```
Comment (FabianWilms) [inline src/components/Form/MucInput.stories.ts:5]:
```suggestion
  title: "Forms/MucInput",
```

File: src/components/Form/MucInput.vue [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        *
+        <span class="visually-hidden">(erforderlich)</span>
+      </span>
+    </label>
+    <p
+      id="text-input-error"
+      class="m-error-message"
+    >
+      {{ errorMsg }}
+    </p>
```
Comment (FabianWilms) [inline src/components/Form/MucInput.vue:25]: Should we hide this if no errorMsg is set?

File: src/components/Form/MucTextArea.stories.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import MucTextArea from "./MucTextArea.vue";
+
+export default {
+  component: MucTextArea,
+  title: "Input/MucTextArea",
```
Comment (FabianWilms) [inline src/components/Form/MucTextArea.stories.ts:5]:
```suggestion
  title: "Forms/MucTextArea",
```

File: src/components/Form/MucTextArea.vue [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+        *
+        <span class="visually-hidden">(erforderlich)</span>
+      </span>
+    </label>
+    <p
+      id="textarea-input-error"
+      class="m-error-message"
+    >
+      {{ errorMsg }}
+    </p>
```
Comment (FabianWilms) [inline src/components/Form/MucTextArea.vue:22]: Same as above

File: src/components/Form/index.ts [hunk 10]
```diff
@@ -0,0 +1,3 @@
+import MucForm from "./MucInput.vue";
+
+export { MucForm };
```
Comment (FabianWilms) [inline src/components/Form/index.ts:3]: here are some imports missing for the other components

**Review by FabianWilms — CHANGES_REQUESTED** [review]

File: src/components/Form/MucCheckbox.vue (hunk 3 above)
Comment (langehm) [reply src/components/Form/MucCheckbox.vue:22]: But [defineModel()](https://vuejs.org/guide/components/v-model#basic-usage) can be accessed from the outside like a prop + does not need any emits to be updated. Maybe i understand something wrong.

File: src/components/Form/MucInput.vue (hunk 7 above)
Comment (langehm) [reply src/components/Form/MucInput.vue:25]: Thought about it, but it does not take up any space on page ... one less v-if?

Comment (FabianWilms) [reply src/components/Form/MucInput.vue:25]: Then we can leave it as is 👍

File: src/components/Form/MucCheckbox.vue (hunk 3 above)
Comment (FabianWilms) [reply src/components/Form/MucCheckbox.vue:22]: > Starting in Vue 3.4, the recommended approach to achieve this is using the [defineModel()](https://vuejs.org/api/sfc-script-setup#definemodel) macro:

sorry, my bad. js libs sometimes move to fast.

Comment (langehm) [reply src/components/Form/MucCheckbox.vue:22]: It's an awesome feature to save some boilerplate code - very much appreciated ☺

File: src/components/Form/MucCheckboxGroup.stories.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+  component: MucCheckboxGroup,
+  title: "Forms/MucCheckboxGroup",
+  tags: ["autodocs"],
+  parameters: {
+    docs: {
+      description: {
+        component: `The muc-checkbox-group component is a wrapper for a list of checkboxes.
+It comes in a collapsable and non-collapsable variant.
+
+[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=viewall-elements-forms)
```
Comment (FabianWilms) [inline src/components/Form/MucCheckboxGroup.stories.ts:15]:
```suggestion
[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=elements-checkboxes-collapse)
```

File: src/components/Form/MucCheckbox.stories.ts [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+  component: MucCheckbox,
+  title: "Forms/MucCheckBox",
+  tags: ["autodocs"],
+  parameters: {
+    docs: {
+      description: {
+        component: `The muc-checkbox component provides the ability to choose between two distinct values.
+It is very similar to a switch and can be used in complex forms.
+
+[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=viewall-elements-forms)
```
Comment (FabianWilms) [inline src/components/Form/MucCheckbox.stories.ts:13]:
```suggestion
[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=elements-checkboxes)
```

File: src/components/Form/MucInput.stories.ts [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+export default {
+  component: MucInput,
+  title: "Forms/MucInput",
+  tags: ["autodocs"],
+  parameters: {
+    docs: {
+      description: {
+        component: `The muc-input component is a text-based input field. It supports various styling options like prefixes /suffixes, buttons and can also be displayed as a textarea.
+
+[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=viewall-elements-forms)
```
Comment (FabianWilms) [inline src/components/Form/MucInput.stories.ts:12]:
```suggestion
[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=elements-input-text)
```

File: src/components/Form/MucRadioButton.stories.ts [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+  title: "Forms/RadioButton",
+  tags: ["autodocs"],
+  parameters: {
+    docs: {
+      description: {
+        component: `
+Offers simple functionality of a radio-button.
+You need to use the radio-button within the radio-button-group - combined you can provide grouping functionality to allow users to select from a predefined set of options.
+
+[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=viewall-elements-forms)
```
Comment (FabianWilms) [inline src/components/Form/MucRadioButton.stories.ts:15]:
```suggestion
[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=elements-radios)
```

File: src/components/Form/MucSelect.stories.ts [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+  title: "Forms/MucSelect",
+  tags: ["autodocs"],
+  parameters: {
+    docs: {
+      description: {
+        component: `
+Select an option from a given list. The design is simmilar to the text input.
+Only one option can be choosen.
+
+[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=viewall-elements-forms)
```
Comment (FabianWilms) [inline src/components/Form/MucSelect.stories.ts:14]:
```suggestion
[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=elements-combobox)
```

File: src/components/Form/RadioButtonGroup.stories.ts [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+import MucRadioButtonGroup from "./MucRadioButtonGroup.vue";
+
+export default {
+  component: MucRadioButtonGroup,
+  title: "Forms/MucRadioButtonGroup",
+  tags: ["autodocs"],
+  parameters: {
+    docs: {
+      description: {
+        component: `The muc-checkbox-group component is a wrapper for a list of checkboxes.
```
Comment (FabianWilms) [inline src/components/Form/RadioButtonGroup.stories.ts:11]: wrong comment?

File: src/components/Form/RadioButtonGroup.stories.ts [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+It comes in a collapsable and non-collapsable variant.
+
+[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=viewall-elements-forms)
```
Comment (FabianWilms) [inline src/components/Form/RadioButtonGroup.stories.ts:14]:
```suggestion
[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=elements-radios)
```

**Comment by langehm** [comment]
We should introduce a composable to define unique id's automagically to every html element. Ussage like the following:
```
const idForThisComponent = useUniqueID("muc-radio-button"); // return sth like muc-radio-button-123
```

vuetify uses similiar approach with some vue methods to generate the unique id's.

**Comment by FabianWilms** [comment]
> We should introduce a composable to define unique id's automagically to every html element. Ussage like the following:
> 
> ```
> const idForThisComponent = useUniqueID("muc-radio-button"); // return sth like muc-radio-button-123
> ```
> 
> vuetify uses similiar approach with some vue methods to generate the unique id's.

https://github.com/it-at-m/muc-patternlab-vue/issues/149

**Comment by langehm** [comment]
Above comment suggestions were done in this [commit](https://github.com/it-at-m/muc-patternlab-vue/pull/101/commits/fc5b52f17af19a9109909c8232976bef49d6c2f3)

File: src/components/Form/MucErrorList.stories.ts [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+  title: "Forms/MucErrorList",
+  tags: ["autodocs"],
+  parameters: {
+    docs: {
+      description: {
+        component: `The muc-checkbox component provides the ability to choose between two distinct values.
+It is very similar to a switch and can be used in complex forms.
+
+[🔗 Patternlab-Docs](https://patternlab.muenchen.space/?p=elements-checkboxes)
+`,
```
Comment (FabianWilms) [inline src/components/Form/MucErrorList.stories.ts:14]: Wrong description for element?

File: src/components/Form/MucRadioButton.stories.ts [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+export default {
+  component: MucRadioButton,
+  title: "Forms/RadioButton",
+  tags: ["autodocs"],
+  parameters: {
+    docs: {
+      description: {
+        component: `
+Offers simple functionality of a radio-button.
+You need to use the radio-button within the radio-button-group - combined you can provide grouping functionality to allow users to select from a predefined set of options.
```
Comment (FabianWilms) [inline src/components/Form/MucRadioButton.stories.ts:13]: maybe add a link to the radio-button-group-story?

**Review by FabianWilms — CHANGES_REQUESTED** [review]
And one small edition when testing out the components: I can select a radio button by clicking on the text, but for the checkbox-component I can only select it by clicking into the box itself. I think clicking the text should also select it.

File: src/components/Form/MucRadioButton.stories.ts (hunk 19 above)
Comment (FabianWilms) [reply src/components/Form/MucRadioButton.stories.ts:13]: sorry, my bad!

**Review by FabianWilms — DISMISSED** [review]

**Pull request merged**